src/pybind/mgr/dashboard/controllers/__init__.py

```diff
@@ -454,146 +544,114 @@ class RESTController(BaseController):
+        return None
 
-        return result
-
-    @cherrypy.expose
-    def _collection(self, *vpath, **params):
-        return self._rest_request(False, *vpath, **params)
-
-    @cherrypy.expose
-    def _element(self, *vpath, **params):
-        return self._rest_request(True, *vpath, **params)
-
-    def _handle_detail_method(self, *vpath, **params):
-        method = getattr(self, cherrypy.request.path_info.split('/')[-1])
-
-        if cherrypy.request.method not in ['GET', 'DELETE']:
-            method = RESTController._takes_json(method)
-
-        method = RESTController._returns_json(method)
-
-        cherrypy.response.status = 200
-
-        return method(*vpath, **params)
+    @classmethod
+    def endpoints(cls):
+        result = super(RESTController, cls).endpoints()
+        for _, func in inspect.getmembers(cls, predicate=callable):
+            no_resource_id_params = False
+            status = 200
+            method = None
+            path = ""
+
+            if func.__name__ in cls._method_mapping:
+                meth = cls._method_mapping[func.__name__]
+
+                if meth['resource']:
+                    res_id_params = cls.infer_resource_id()
+                    if res_id_params is None:
+                        no_resource_id_params = True
+                    else:
+                        res_id_params = ["{{{}}}".format(p) for p in res_id_params]
+                        path += "/{}".format("/".join(res_id_params))
 
-    def _rest_request(self, is_element, *vpath, **params):
-        method_name, status_code = self._method_mapping[
-            (cherrypy.request.method, is_element)]
-        method = getattr(self, method_name, None)
+                status = meth['status']
+                method = meth['method']
 
-        if cherrypy.request.method not in ['GET', 'DELETE']:
-            method = RESTController._takes_json(method)
+            elif hasattr(func, "_collection_method_"):
+                path = "/{}".format(func.__name__)
+                method = func._collection_method_
 
-        method = RESTController._returns_json(method)
+            elif hasattr(func, "_resource_method_"):
+                res_id_params = cls.infer_resource_id()
+                if res_id_params is None:
+                    no_resource_id_params = True
+                else:
+                    res_id_params = ["{{{}}}".format(p) for p in res_id_params]
+                    path += "/{}".format("/".join(res_id_params))
+                    path += "/{}".format(func.__name__)
 
-        cherrypy.response.status = status_code
+                method = func._resource_method_
 
-        return method(*vpath, **params)
+            else:
+                continue
 
-    @staticmethod
-    def _function_args(func):
-        return getargspec(func).args[1:]
+            if no_resource_id_params:
+                raise TypeError("Invalid resource method '{}'. Please check if "
```

Comment:
I still do not understand what this message is telling me. Especially the part _be used has resource id, or specify the_. Maybe a comma is missing or rearrange the sentence to clarify the meaning.